src/Plugin/Commerce/PaymentGateway/Heartland.php

```diff
@@ -0,0 +1,391 @@
+                ->void()
+                ->execute();
+  
+            // Commerce plugin sets state for next step
+            $payment->setState('authorization_voided');
+            $payment->save();
+        } catch (GatewayException $e) {
+            // If void fails because the auth/capture has completed and the batch is closed/settled
+            // Do a refund instead
+            print_r('Void Failed, Try to Refund<br/>');
+            $this->refundPayment($payment, $amount);
+        }
+    }
+  
+    /**
+     * {@inheritdoc}
+     */
+    public function refundPayment(PaymentInterface $payment, Price $amount = null)
+    {
+        $this->assertPaymentState($payment, ['completed', 'partially_refunded']);
+        // If not specified, refund the entire amount.
+        $amount = $amount ?: $payment->getAmount();
+        $this->assertRefundAmount($payment, $amount);
+
+        // Perform the refund request here, throw an exception if it fails.
+        // See \Drupal\commerce_payment\Exception for the available exceptions.
+
+        // Set some variables to use for our transaction
+        $remote_id = $payment->getRemoteId();
+        $number = $amount->getNumber();
+        $currency = $amount->getCurrencyCode();
+
+        // Heartland SDK - Refund Transaction
+        // https://developer.heartlandpaymentsystems.com/Documentation/credit-card-payments/#refund-a-transaction
+        try {
+            $response = Transaction::fromId($remote_id, PaymentMethodType::CREDIT)
+                ->refund($number)
+                ->withCurrency($currency)
+                ->execute();
+  
+            // In case more than one refund has occurred, Commerce plugin sets amounts and states
+            $old_refunded_amount = $payment->getRefundedAmount();
+            $new_refunded_amount = $old_refunded_amount->add($amount);
+            if ($new_refunded_amount->lessThan($payment->getAmount())) {
+                $payment->setState('partially_refunded');
+            } else {
+                $payment->setState('refunded');
+            }
+    
+            $payment->setRefundedAmount($new_refunded_amount);
+            $payment->save();
+        } catch (GatewayException $e) {
+            throw new PaymentGatewayException($e->getMessage());
+        }
+    }
+  
+    /**
+     * {@inheritdoc}
+     */
+    public function createPaymentMethod(PaymentMethodInterface $payment_method, array $payment_details)
+    {
+        $required_keys = [
+            // The expected keys are payment gateway specific and usually match
+            // the PaymentMethodAddForm form elements. They are expected to be valid.
+            'token_value',
+        ];
+        foreach ($required_keys as $required_key) {
+            if (empty($payment_details[$required_key])) {
+                throw new \InvalidArgumentException(sprintf('$payment_details must contain the %s key.', $required_key));
```

Comment:
Does this exception message bubble up to a user? If so, can we make this more user friendly? `$payment_details` probably won't mean much to normal users.

Reply:
This message only appears if verbose debugging is on.  The user simply sees a very generic site error message.